polyglot/piranha/rules_dot_graph.py

```diff
@@ -0,0 +1,177 @@
+
+# REQUIRED: pip install toml
+
+# the graph can be very wide if there are several disconnected rules
+# Here are a few options:
+# https://stackoverflow.com/questions/11134946/distribute-nodes-on-the-same-rank-of-a-wide-graph-to-different-lines
+# unflatten is probably the best option but requires a graphviz installation
+# as suggested by the SO answer, one can add invisible edges (in the generated .dot) between disconnected nodes
+# Example:
+# python rules_dot_graph.py src/cleanup_rules/kt test-resources/kt/feature_flag_system_1/control/configurations  --title="Kotlin Test Feature Flag System 1" > temp.dot
+# Then in the .dot file:
+# edge[style=invis];
+# delete_all_statements_after_return -> delete_putToggleEnabled
+# delete_putToggleEnabled -> delete_putToggleDisabled
+# delete_putToggleDisabled -> delete_includeEvent
+# edge[style=solid];
+# // all the generated edges
+
+
+@dataclass(frozen=True, eq=True)
+class Edge:
+    to: str
+    scope: str
+
+
+example_text = '''For a complete output, the script needs the directory with the built-in rules for a given language. Example:
+
+    python rules_dot_graph.py src/cleanup_rules/java demo/feature_flag_cleanup/java/configurations --title "Feature Flag Cleanup"'''
+
+description_text = '''Script to output a .dot graph for visualizing how rules/groups are connected.
+Please install the `toml` PyPi package: `python install toml`.
+The script assumes that rules will have at [0,2] groups. If a rule has two, one will be `Cleanup Rule`. The visualization will likely break if any rule has > 2 groups.'''
+
+arg_parser = argparse.ArgumentParser(
+    description=description_text, epilog=example_text, formatter_class=argparse.RawDescriptionHelpFormatter)
+arg_parser.add_argument('configurations_path', type=str, nargs='+',
+                        help="One or more root directory that contains 'rules.toml' and 'edges.toml'")
+arg_parser.add_argument('--title', nargs='?',
+                        help='Optional title for the graph')
+args = arg_parser.parse_args()
+
+
+def sanitize_name(s: str) -> str:
+    """Graphviz does not like names with spaces. Converts spaces to '_'."""
+    s = re.sub(r"\s+", '_', s)
+    return s
+
+
+# dict's are ordered
+rules_by_group_dict: 'dict[str, list[str]]' = {}
+outgoing_edges_by_node: 'dict[str, list[Edge]]' = {}
+
+cleanup_rules: 'set[str]' = set()
+orphan_nodes: 'set[str]' = set()
+
+for config_path in args.configurations_path:
+    config_path = os.path.abspath(config_path) + os.path.sep
+    rules_file = config_path + 'rules.toml'
+
+    # rules.toml should exist, deliberately fail otherwise
+    rules_toml_dict = toml.load(rules_file)
+
+    for rule_toml in rules_toml_dict['rules']:
+        rule_name: str = sanitize_name(rule_toml['name'])
+        if 'groups' in rule_toml:
+            for group_name in rule_toml['groups']:
+                group_name = sanitize_name(group_name)
+                if group_name != 'Cleanup_Rule':
```

Comment:
i think there is no `_`

Reply:
You're right. In `rules.toml`, and probably in the rust code, it's `Cleanup Rule`.
For Graphviz, we are converting all blank spaces to `_`. The conversion makes our lives easier when defining nodes and edges. I'm not sure Graphviz plays well with spaces in names.

https://github.com/uber/piranha/pull/262/files/5b26d2d4c3dae196cf76376509182a11e7e9cd1f#diff-62d9897c4603bf7c92c8ea08ac5a69b1cb508aa8f9cf262bac0e3d83649fc425R48-R51

Reply:
I see. I see. Could be a good naming convention to add .